onnxruntime/core/framework/allocation_planner.cc

```diff
@@ -572,41 +580,125 @@ class PlannerImpl {
+    for (const auto& node : graph_viewer.Nodes()) {
+      const auto& input_node_args = node.InputDefs();
+      size_t num_node_inputs = input_node_args.size();
+
+      for (size_t node_input_index = 0; node_input_index < num_node_inputs; ++node_input_index) {
+        auto input_node_arg = input_node_args[node_input_index];
+
+        // Skip processing missing optional inputs
+        if (!input_node_arg->Exists()) {
+          continue;
+        }
+
+        ORT_TRY {
+          auto& def_name = input_node_arg->Name();
+
+          // This node input doesn't correspond to any of the weights
+          if (!weights.count(def_name)) {
+            continue;
+          }
+
+          // Skip processing shadow values in subgraphs.
+          // While processing subgraphs, if we don't see this entry in the implicit
+          // inputs of the node containing the subgraph, skip processing it as it is
+          // a shadow value.
+          if (graph_depth > 0) {
+            // We are processing a subgraph if we enter this
+            const auto* parent_node = graph_viewer.ParentNode();
+            ORT_ENFORCE(parent_node);
+
+            bool is_implicit_input = false;
+            for (const auto& implicit_input : parent_node->ImplicitInputDefs()) {
+              if (implicit_input->Name() == def_name) {
+                is_implicit_input = true;
+                break;
+              }
             }
-            ORT_CATCH(const std::exception& ex) {
-              ORT_HANDLE_EXCEPTION([&]() {
-                sub_status = ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, ex.what());
-              });
+
+            // Only proceed if it is an implicit input
+            if (!is_implicit_input) {
+              continue;
             }
-            return sub_status;
+          }
+
+          auto wt_index = Index(def_name);
+          locations[wt_index].emplace_back(GetLocationForNodeInput(node_input_index, node, kernel_create_info_map, execution_providers));
+        }
+        ORT_CATCH(const std::exception& ex) {
+          ORT_HANDLE_EXCEPTION([&]() {
+            return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, ex.what());
           });
+        }
+      }
 
-      ORT_RETURN_IF_ERROR(status);
+      // If the node has subgraphs (i.e.) control flow nodes,
+      // walk the nodes in those subgraphs as well to best determine
+      // the location for the OrtValue corresponding to the weights
+      // (i.e.) do a recursion
+      if (node.ContainsSubgraph()) {
+        for (auto& name_to_subgraph : node.GetAttributeNameToSubgraphMap()) {
+          GraphViewer subgraph_viewer(*name_to_subgraph.second);
+
+          std::ostringstream ss;
+
+          // key = base + depth + current graph node index + attr name corresponding to the subgraph
+          ss << subgraph_kernel_create_info_map_key_base;
+          ss << graph_depth;
+          ss << node.Index();
+          ss << name_to_subgraph.first;
+
```

Comment:
If this needs to match the key logic in session_state.cc there should be a shared helper for it.